contrib/platform/src/com/sun/jna/platform/win32/WinspoolUtil.java

```diff
@@ -135,28 +88,83 @@ public static PRINTER_INFO_2 getPrinterInfo2(String printerName) {
             throw we;
         }
 
-        return pinfo2;
+        return struct;
     }
 
-    public static PRINTER_INFO_4[] getPrinterInfo4() {
+    /**
+     * Helper for getting array of printer info
+     * PRINTER_ENUM_LOCAL | PRINTER_ENUM_CONNECTIONS
+     */
+    static Structure[] getPrinterInfoByStruct(int flags, int structType) {
         IntByReference pcbNeeded = new IntByReference();
         IntByReference pcReturned = new IntByReference();
-        Winspool.INSTANCE.EnumPrinters(Winspool.PRINTER_ENUM_LOCAL, null, 4,
-                null, 0, pcbNeeded, pcReturned);
+        // When Name is NULL, setting Flags to PRINTER_ENUM_LOCAL | PRINTER_ENUM_CONNECTIONS
+        // enumerates printers that are installed on the local machine.
+        // These printers include those that are physically attached to the local machine
+        // as well as remote printers to which it has a network connection.
+        // See https://msdn.microsoft.com/en-us/library/windows/desktop/dd162692(v=vs.85).aspx
+        Winspool.INSTANCE.EnumPrinters(flags, null, structType, null, 0, pcbNeeded, pcReturned);
+        Structure struct = initStructByType(structType, pcbNeeded.getValue());
         if (pcbNeeded.getValue() <= 0) {
-            return new PRINTER_INFO_4[0];
+            return emptyStructArrayByType(structType);
         }
-
-        PRINTER_INFO_4 pPrinterEnum = new PRINTER_INFO_4(pcbNeeded.getValue());
-        if (!Winspool.INSTANCE.EnumPrinters(Winspool.PRINTER_ENUM_LOCAL, null,
-                4, pPrinterEnum.getPointer(), pcbNeeded.getValue(), pcbNeeded,
+        if (!Winspool.INSTANCE.EnumPrinters(flags, null, structType, struct.getPointer(), pcbNeeded.getValue(), pcbNeeded,
                 pcReturned)) {
             throw new Win32Exception(Kernel32.INSTANCE.GetLastError());
         }
 
-        pPrinterEnum.read();
+        struct.read();
+        return struct.toArray(pcReturned.getValue());
+    }
+
+    private static Structure[] emptyStructArrayByType(int structType) {
```

Comment:
Ironically, the underlying Winspool APIs uses integers for the struct types, however, ince Java is strongly typed, we have to write some helpers to make sure we return the structs we asked for.

The old API would return an empty struct when searching for a single printer, but an empty array when searching for all printers.  This is why `initStructByType` and `emptyStructArrayByType` needed two separate functions.

If the names of these helpers are not intuitive, advice is welcome. :)